liquidator/liquidator.js

```diff
@@ -158,12 +168,69 @@ class Liquidator {
       // Note: query the time again during each iteration to ensure the deadline is set reasonably.
       const currentBlockTime = this.empClient.getLastUpdateTime();
 
+      // Calculate the amount of tokens we will attempt to liquidate.
+      let tokensToLiquidate;
+
+      // If the user specifies `maxTokensToLiquidateWei`, then we must make sure that it follows an important constraint:
+      // we cannot bring the position down below the `minSponsorPosition`.
+      if (maxTokensToLiquidateWei) {
+        // First, we check if `maxTokensToLiquidateWei > tokensOutstanding`, if so then we'll liquidate the full position.
+        if (this.toBN(maxTokensToLiquidateWei).gte(this.toBN(position.numTokens))) {
+          tokensToLiquidate = this.toBN(position.numTokens);
+        } else {
+          // If we're not liquidating the full position, then we cannot liquidate the position below the `minSponsorTokens` constraint.
+          // `positionTokensAboveMinimum` is the maximum amount of tokens any liquidator can liquidate while taking `minSponsorTokens` into account.
+          const positionTokensAboveMinimum = this.toBN(position.numTokens).sub(this.toBN(this.empMinSponsorSize));
+
+          // Finally, we cannot liquidate more than `maxTokensToLiquidate`.
+          tokensToLiquidate = this.BN.min(positionTokensAboveMinimum, this.toBN(maxTokensToLiquidateWei));
+        }
+      } else {
+        // If `maxTokensToLiquidateWei` is not specified, then we will attempt to liquidate the full position.
+        tokensToLiquidate = this.toBN(position.numTokens);
+      }
+
+      // If `tokensToLiquidate` is 0, then skip this liquidation. Due to the if-statement branching above, `tokensToLiquidate == 0`
+      // is only possible if the `positionTokensAboveMinimum == 0` && `maxTokensToLiquidate < position.numTokens`. In other words,
+      // the bot cannot liquidate the full position size, but the full position size is at the minimum sponsor threshold. Therefore, the
+      // bot can liquidate 0 tokens. The smart contracts should disallow this, but a/o June 2020 this behavior is allowed so we should block it
+      // client-side.
+      if (tokensToLiquidate.isZero()) {
+        this.logger.error({
+          at: "Liquidator",
+          message: "Position size is equal to the minimum: not enough synthetic to initiate full liquidation✋",
+          sponsor: position.sponsor,
+          inputPrice: scaledPrice.toString(),
+          position: position,
+          minLiquidationPrice: this.liquidationMinPrice,
+          maxLiquidationPrice: maxCollateralPerToken.toString(),
+          tokensToLiquidate: tokensToLiquidate.toString(),
+          error: new Error("Refusing to liquidate 0 tokens")
+        });
+        continue;
+      }
+
+      // Send an alert if the bot is going to submit a partial liquidation instead of a full liquidation.
+      if (tokensToLiquidate.lt(this.toBN(position.numTokens))) {
```

Comment:
Error message 2: Position size > minimum, but liq bot balance < position size. Alerting that we are sending a partial liquidation